File: android/src/main/java/com/wortise/ads/react/RNWortiseNativeAdLoader.kt
package com.wortise.ads.react

import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.LifecycleEventListener
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.bridge.WritableMap
import com.facebook.react.modules.core.DeviceEventManagerModule
import com.wortise.ads.AdError
import com.wortise.ads.RevenueData
import com.wortise.ads.natives.NativeAd
import com.wortise.ads.natives.NativeAdLoader
import com.wortise.ads.react.extensions.toRequestParameters
import com.wortise.ads.react.extensions.toWritableMap
import java.util.UUID

class RNWortiseNativeAdLoader(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext), LifecycleEventListener {

    private val eventEmitter by lazy {
        reactContext.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
    }


    init {
        reactContext.addLifecycleEventListener(this)
    }


    override fun getName(): String = "RNWortiseNativeAdLoader"

    @ReactMethod
    fun addListener(eventName: String) {}

    @ReactMethod
    fun removeListeners(count: Int) {}

    @ReactMethod
    fun loadAd(adUnitId: String, requestParameters: ReadableMap?, promise: Promise) {
        val context = reactApplicationContext.currentActivity ?: reactApplicationContext

        val responseId = UUID.randomUUID().toString()

        val listener = object : NativeAdLoader.Listener {

            override fun onNativeLoaded(ad: NativeAd) {
                loadedAds[responseId] = ad

                promise.resolve(serializeNativeAd(responseId, ad))
            }

            override fun onNativeFailedToLoad(error: AdError) {
                adLoaders.remove(responseId)?.destroy()

                promise.reject("AD_LOAD_FAILED", error.toString())
            }

            override fun onNativeClicked(ad: NativeAd) {
                val params = Arguments.createMap()

                params.putString("responseId", responseId)

                sendEvent(EVENT_CLICKED, params)
            }

            override fun onNativeImpression(ad: NativeAd) {
                val params = Arguments.createMap()

                params.putString("responseId", responseId)

                sendEvent(EVENT_IMPRESSION, params)
            }

            override fun onNativeRevenuePaid(ad: NativeAd, data: RevenueData) {
                val params = Arguments.createMap()

                params.putString("responseId", responseId)
                params.putMap   ("data",       data.toWritableMap())

                sendEvent(EVENT_REVENUE_PAID, params)
            }
        }

        val loader = NativeAdLoader(context, adUnitId, listener)

        adLoaders[responseId] = loader

        val parameters = requestParameters.toRequestParameters()

        loader.loadAd(parameters)
    }

    @ReactMethod
    fun destroyAd(responseId: String) {
        loadedAds.remove(responseId)
        adLoaders.remove(responseId)?.destroy()
    }


    private fun serializeNativeAd(responseId: String, ad: NativeAd): WritableMap {
        val map = Arguments.createMap()

        map.putString("responseId",   responseId)
        map.putString("headline",     ad.headline)
        map.putString("body",         ad.body)
        map.putString("callToAction", ad.callToAction)
        map.putString("advertiser",   ad.advertiser)
        map.putString("price",        ad.price)
        map.putString("store",        ad.store)

        ad.rating?.let { map.putDouble("rating", it) }

        ad.icon?.let { icon ->
            val iconMap = Arguments.createMap()

            icon.uri?.let { iconMap.putString("uri", it.toString()) }

            icon.drawable?.let { drawable ->
                iconMap.putInt("width",  drawable.intrinsicWidth)
                iconMap.putInt("height", drawable.intrinsicHeight)
            }

            map.putMap("icon", iconMap)
        }

        val images = ad.images

        if (images.isNotEmpty()) {
            val imagesArray = Arguments.createArray()

            for (image in images) {
                val imageMap = Arguments.createMap()

                image.uri?.let { imageMap.putString("uri", it.toString()) }

                image.drawable?.let { drawable ->
                    imageMap.putInt("width",  drawable.intrinsicWidth)
                    imageMap.putInt("height", drawable.intrinsicHeight)
                }

                imagesArray.pushMap(imageMap)
            }

            map.putArray("images", imagesArray)
        }

        ad.mediaContent?.let { media ->
            val mediaMap = Arguments.createMap()

            media.aspectRatio?.let { mediaMap.putDouble("aspectRatio", it.toDouble()) }

            map.putMap("mediaContent", mediaMap)
        }

        return map
    }

    private fun sendEvent(eventName: String, params: WritableMap? = null) {
        eventEmitter.emit(eventName, params)
    }


    override fun onHostDestroy() {
        loadedAds.clear()

        adLoaders.values.forEach { it.destroy() }
        adLoaders.clear()
    }

    override fun onHostPause() {}

    override fun onHostResume() {}


    companion object {
        const val EVENT_CLICKED      = "onNativeClicked"
        const val EVENT_IMPRESSION   = "onNativeImpression"
        const val EVENT_REVENUE_PAID = "onNativeRevenuePaid"


        val adLoaders = mutableMapOf<String, NativeAdLoader>()

        val loadedAds = mutableMapOf<String, NativeAd>()
    }
}

File: ios/RNWortiseSdk/RNWortiseNativeAdLoader.swift
import Foundation
import React
import WortiseSDK

@objc(RNWortiseNativeAdLoader)
class RNWortiseNativeAdLoader: RCTEventEmitter {

    fileprivate static let EVENT_CLICKED      = "onNativeClicked"
    fileprivate static let EVENT_IMPRESSION   = "onNativeImpression"
    fileprivate static let EVENT_REVENUE_PAID = "onNativeRevenuePaid"

    static var sharedAds     = [String: WANativeAd]()
    static var sharedLoaders = [String: WANativeAdLoader]()


    fileprivate var delegates = [String: NativeAdDelegateWrapper]()


    override static func requiresMainQueueSetup() -> Bool {
        return false
    }

    override func supportedEvents() -> [String]! {
        return [
            RNWortiseNativeAdLoader.EVENT_CLICKED,
            RNWortiseNativeAdLoader.EVENT_IMPRESSION,
            RNWortiseNativeAdLoader.EVENT_REVENUE_PAID
        ]
    }


    @objc(destroyAd:)
    func destroyAd(_ responseId: String) {
        RNWortiseNativeAdLoader.sharedAds.removeValue(forKey: responseId)
        RNWortiseNativeAdLoader.sharedLoaders.removeValue(forKey: responseId)?.destroy()

        delegates.removeValue(forKey: responseId)
    }

    @objc(loadAd:requestParameters:resolve:reject:)
    func loadAd(_ adUnitId:        String,
                requestParameters: NSDictionary?,
                resolve:           @escaping RCTPromiseResolveBlock,
                reject:            @escaping RCTPromiseRejectBlock) {

        let responseId = UUID().uuidString

        DispatchQueue.main.async { [weak self] in

            guard let self = self else { return }

            let delegate = NativeAdDelegateWrapper(
                responseId: responseId,
                resolve: resolve,
                reject: reject,
                owner: self
            )

            let loader = WANativeAdLoader(adUnitId: adUnitId, delegate: delegate)

            self.delegates[responseId] = delegate

            RNWortiseNativeAdLoader.sharedLoaders[responseId] = loader

            let parameters = WARequestParameters.from(requestParameters)

            loader.loadAd(parameters: parameters)
        }
    }


    fileprivate func serializeNativeAd(_ responseId: String, ad: WANativeAd) -> [String: Any?] {
        var map: [String: Any?] = [
            "advertiser":   ad.advertiser,
            "body":         ad.body,
            "callToAction": ad.callToAction,
            "headline":     ad.headline,
            "price":        ad.price,
            "responseId":   responseId,
            "store":        ad.store
        ]

        if let rating = ad.rating {
            map["rating"] = rating
        }

        if let icon = ad.icon {

            var iconMap: [String: Any?] = [:]

            iconMap["uri"] = icon.url?.absoluteString

            if let image = icon.image {
                iconMap["width"]  = Int(image.size.width)
                iconMap["height"] = Int(image.size.height)
                iconMap["scale"]  = image.scale
            }

            map["icon"] = iconMap
        }

        let images = ad.images

        if !images.isEmpty {

            var imagesArray: [[String: Any?]] = []

            for nativeImage in images {

                var imageMap: [String: Any?] = [:]

                imageMap["uri"] = nativeImage.url?.absoluteString

                if let img = nativeImage.image {
                    imageMap["width"]  = Int(img.size.width)
                    imageMap["height"] = Int(img.size.height)
                    imageMap["scale"]  = img.scale
                }

                imagesArray.append(imageMap)
            }

            map["images"] = imagesArray
        }

        if let media = ad.mediaContent {

            var mediaMap: [String: Any?] = [:]

            if let aspectRatio = media.aspectRatio {
                mediaMap["aspectRatio"] = aspectRatio
            }

            map["mediaContent"] = mediaMap
        }

        return map
    }

    fileprivate func storeAd(_ responseId: String, ad: WANativeAd) {
        RNWortiseNativeAdLoader.sharedAds[responseId] = ad
    }
}

private class NativeAdDelegateWrapper: NSObject, WANativeDelegate {

    let responseId: String

    var resolve: RCTPromiseResolveBlock?
    var reject:  RCTPromiseRejectBlock?

    weak var owner: RNWortiseNativeAdLoader?


    init(responseId: String,
         resolve:    @escaping RCTPromiseResolveBlock,
         reject:     @escaping RCTPromiseRejectBlock,
         owner:      RNWortiseNativeAdLoader) {
        self.responseId = responseId
        self.resolve    = resolve
        self.reject     = reject
        self.owner      = owner
    }


    func didClick(nativeAd: WANativeAd) {
        let body = [
            "responseId": responseId
        ]

        owner?.sendEvent(
            withName: RNWortiseNativeAdLoader.EVENT_CLICKED,
            body: body
        )
    }

    func didFailToLoad(nativeAd error: WAAdError) {
        RNWortiseNativeAdLoader.sharedLoaders.removeValue(forKey: responseId)?.destroy()

        reject?("AD_LOAD_FAILED", error.message, nil)

        resolve = nil
        reject  = nil
    }

    func didImpress(nativeAd: WANativeAd) {
        let body = [
            "responseId": responseId
        ]

        owner?.sendEvent(
            withName: RNWortiseNativeAdLoader.EVENT_IMPRESSION,
            body: body
        )
    }

    func didLoad(nativeAd: WANativeAd) {
        owner?.storeAd(responseId, ad: nativeAd)

        let data = owner?.serializeNativeAd(responseId, ad: nativeAd)

        resolve?(data)

        resolve = nil
        reject  = nil
    }

    func didPayRevenue(nativeAd: WANativeAd, data: WARevenueData) {
        let body: [String : Any] = [
            "responseId": responseId,
            "data": data.toMap()
        ]

        owner?.sendEvent(
            withName: RNWortiseNativeAdLoader.EVENT_REVENUE_PAID,
            body: body
        )
    }
}
